### PR TITLE
Parser2 support for while and expressions

### PR DESCRIPTION
**Pull request by jonmeow**
Adds `while` and most of the expression support. Splits apart the fixity test to demonstrate more closely which bits are still failing.

```
//toolchain/parser/testdata:basics/fail_paren_match_regression.carbon.test FAILED in 0.8s
//toolchain/parser/testdata:basics/function_call.carbon.test             FAILED in 0.7s
//toolchain/parser/testdata:basics/package.carbon.test                   FAILED in 0.7s
//toolchain/parser/testdata:basics/structs.carbon.test                   FAILED in 0.7s
//toolchain/parser/testdata:basics/tuples.carbon.test                    FAILED in 0.6s
//toolchain/parser/testdata:basics/var.carbon.test                       FAILED in 0.7s
//toolchain/parser/testdata:for/fail_colon_instead_of_in.carbon.test     FAILED in 0.7s
//toolchain/parser/testdata:for/fail_missing_in.carbon.test              FAILED in 0.7s
//toolchain/parser/testdata:for/fail_missing_var.carbon.test             FAILED in 0.7s
//toolchain/parser/testdata:for/nested.carbon.test                       FAILED in 0.6s
//toolchain/parser/testdata:for/simple.carbon.test                       FAILED in 0.8s
//toolchain/parser/testdata:function/definition/with_params.carbon.test  FAILED in 0.8s
//toolchain/parser/testdata:operators/fixity_in_call.carbon.test         FAILED in 0.7s
//toolchain/parser/testdata:operators/fixity_in_var.carbon.test          FAILED in 0.8s
```

### CODE REVIEW

File: toolchain/parser/parser2.h [hunk 1]
```diff
@@ -38,21 +42,49 @@ class Parser2 {
...
+          token(token),
+          subtree_start(subtree_start) {}
 
     // The state.
     ParserState state;
+    // Set to true  to indicate that an error was found, and that contextual
+    // error recovery may be needed.
+    bool has_error = false;
+
+    // Precedence information used for operator states.
```
Comment (chandlerc) [inline toolchain/parser/parser2.h]: Maybe expand this a bit to help document how these are used?

File: toolchain/parser/testdata/operators/fixity_with_assign.carbon [hunk 2]
```diff
@@ -59,26 +35,13 @@
 // CHECK:STDOUT:       {kind: 'InfixOperator', text: '*', subtree_size: 3},
```
Comment (chandlerc) [inline toolchain/parser/testdata/operators/fixity_with_assign.carbon:35]: Is there a FIXME for the infix here?

File: toolchain/parser/testdata/operators/fixity_with_assign.carbon [hunk 3]
```diff
@@ -59,26 +35,13 @@
...
 // CHECK:STDOUT: ]
 
-fn F(p: i32*, n: i32) {
-  var q: i32* = p;
-  var t: Type = i32*;
+fn F() {
   t = t**;
   n = n * n;
   n = n * *p;
   n = n*n;
```
Comment (chandlerc) [inline toolchain/parser/testdata/operators/fixity_with_assign.carbon:46]: These don't seem to show up in the test? Am I missing something?

**Review by chandlerc — COMMENTED** [review]
Need to read the actual code a bit more carefully, but some quick initial comments.

File: toolchain/parser/testdata/operators/fixity_with_assign.carbon (hunk 2 above)
Comment (jonmeow) [reply toolchain/parser/testdata/operators/fixity_with_assign.carbon:35]: I don't think there's an issue here, can you please clarify what your concern is?

Bear in mind, output in these tests is from the current Parser, not the Parser2.

File: toolchain/parser/testdata/operators/fixity_with_assign.carbon (hunk 3 above)
Comment (jonmeow) [reply toolchain/parser/testdata/operators/fixity_with_assign.carbon:46]: Lines 33-37, which you commented on.

File: toolchain/parser/parser2.cpp [hunk 4]
```diff
@@ -17,6 +17,35 @@
 
 namespace Carbon {
 
+// May be omitted a couple different ways as part of operator parsing.
```
Comment (ergawy) [inline toolchain/parser/parser2.cpp]: `omitted` -> `emitted`?

File: toolchain/parser/parser2.h [hunk 5]
```diff
@@ -161,14 +216,14 @@ class Parser2 {
   // Handles a code block in the context of a statement scope.
   auto HandleCodeBlock() -> void;
 
+  // Resumes processing of an expression, with the recently processed expression
+  // indicated by the passed node_kind.
+  auto HandleExpressionLoop(StateStackEntry state) -> void;
```
Comment (ergawy) [inline toolchain/parser/parser2.h]: I think this is not implemented or used, right?

File: toolchain/parser/parser2.cpp [hunk 6]
```diff
@@ -288,8 +437,73 @@ auto Parser2::HandleDeclarationState() -> void {
...
+      ReturnErrorOnState();
+    }
+  }
+
+  AddNode(ParseNodeKind::DesignatorExpression(), *dot, state.subtree_start,
+          state.has_error);
+}
+
+auto Parser2::HandleExpressionState() -> void {
+  // TODO: This is temporary, we should need this state. If not, maybe add an
```
Comment (ergawy) [inline toolchain/parser/parser2.cpp]: we should _not_ need ...?

**Comment by ergawy** [comment]
Just a couple of trivial comments since I am following the PR and eager to jump back on contributing to the toolchain in its new look :).

File: toolchain/parser/parser2.cpp (hunk 4 above)
Comment (jonmeow) [reply toolchain/parser/parser2.cpp]: Fixing

File: toolchain/parser/parser2.cpp (hunk 6 above)
Comment (jonmeow) [reply toolchain/parser/parser2.cpp]: Removing comment, obsolete TODO

File: toolchain/parser/parser2.h (hunk 1 above)
Comment (jonmeow) [reply toolchain/parser/parser2.h]: Added some comments, trying to guess from code. It's not deeply commented in the current Parser.

File: toolchain/parser/parser2.h (hunk 5 above)
Comment (jonmeow) [reply toolchain/parser/parser2.h]: Removed, this was indeed litter from draft implementations

File: toolchain/parser/testdata/operators/fixity_with_assign.carbon (hunk 2 above)
Comment (chandlerc) [reply toolchain/parser/testdata/operators/fixity_with_assign.carbon:35]: Sorry, the github diff Ui when I was looking at this tricked me into thinking that this `infix` went around the `t**` postfix above. Hiding the other postfix in `t**` ... *and all of the other statements* ...

Anyways, now that I have successfully expanded the diff, I see how this works, and all of this makes sense.

File: toolchain/parser/testdata/operators/fixity_with_assign.carbon (hunk 3 above)
Comment (chandlerc) [reply toolchain/parser/testdata/operators/fixity_with_assign.carbon:46]: Github diff UI issue, see above. Sorry.

File: toolchain/parser/parser2.h [hunk 7]
```diff
@@ -38,21 +42,52 @@ class Parser2 {
...
+          ambient_precedence(ambient_precedence),
+          lhs_precedence(lhs_precedence),
+          token(token),
+          subtree_start(subtree_start) {}
 
     // The state.
     ParserState state;
+    // Set to true  to indicate that an error was found, and that contextual
+    // error recovery may be needed.
+    bool has_error = false;
```
Comment (chandlerc) [inline toolchain/parser/parser2.h]:
```suggestion
    // Set to true to indicate that an error was found, and that contextual
    // error recovery may be needed.
    bool has_error = false;
```

**Review by chandlerc — APPROVED** [review]
LGTM

**Pull request merged**